Added support for transforming homogenous W geometry into world geomtry

dkollmann opened the pull request:
This is an update on the previous pull request https://github.com/elishacloud/dxwrapper/pull/188.

dkollmann committed:
Added support for transforming homogenous W geometry into world geometry.

Trass3r commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -16,9 +16,10 @@
 
 // Enable for testing only
-//#define ENABLE_DEBUGOVERLAY
+#define ENABLE_DEBUGOVERLAY
```
Accidental?

elishacloud replied:
Fixed in the latest check-in.

elishacloud commented:
Thanks @dkollmann!  I will take a look at this soon.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,15 +339,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						const float fov = Config.DdrawConvertHomogeneousToWorldFOV;
+						const float nearplane = Config.DdrawConvertHomogeneousToWorldNearPlane;
+						const float farplane = Config.DdrawConvertHomogeneousToWorldFarPlane;
+						DirectX::XMMATRIX proj = DirectX::XMMatrixPerspectiveFovLH(fov * (3.14159265359f / 180.0f), ratio, nearplane, farplane);
```
This is the only place `ratio` is read, so it could be declared 1 line above, instead of all the way up.

elishacloud replied:
I'm going to leave this.  I am ok with all the constants being declared at the beginning.

dkollmann commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+					{
+						float *pos = (float*) vertex;
+
+						pos[3] = 1.0f;
```
Still not 100% sure if this is even needed. I think it made no difference for Black & White.

Trass3r replied:
Could be they only use this for UI, in which case it probably already is 1.0f.

elishacloud replied:
This has been commented out for now.

dkollmann commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,15 +339,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+					else
+					{
+						// Override the original matrix
+						std::memcpy(lpD3DMatrix, &view, sizeof(_D3DMATRIX));
```
Just realized this basically ended up being the same code as above. So it can be moved out.

elishacloud replied:
Fixed this in the latest check in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+						float *pos = (float*) vertex;
+
+						pos[3] = 1.0f;
+
```
This changes the contents of the source geometry. Generally it's not a good idea to do this, since that memory might reside in read-only memory, might be hashed, or used for other purposes. Such use cases will break. Better use an intermediate copy, just like the else case, except it can be copied over fully before modifications the single component

elishacloud replied:
This has been commented out.

dkollmann commented:
Also FYI if you try this code with Black & White, things will have a blue tint. This is because the water plane which should be rendered before all the other geometry, is rendered on top of it instead of behind the other geometry.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,15 +339,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+
+						DirectX::XMVECTOR up = DirectX::XMVectorSet(0.0f, 1.0f, 0.0f, 0.0f);
+
+						DirectX::XMMATRIX viewMatrix;
```
DirectX::XMVECTOR position and direction could be declared here, assigned in either path of the if, so that viewMatrix could be declared and assigned just once after below if+else blocks.

elishacloud replied:
This has been fixed in the latest check in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+				}
+				else
+				{
+					const UINT newstride = stride - sizeof(float);
```
Perhaps rename `newstride` into `targetStride`?

elishacloud replied:
Fixed in the latest check in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+			const UINT stride = GetVertexStride(dwVertexTypeDesc);
+
+			// Handle PositionT
+			if((dwVertexTypeDesc & D3DFVF_XYZRHW) != 0 && Config.DdrawConvertHomogeneousW)
```
Instead of this, check `(dwVertexTypeDesc & 0x0E) == D3DFVF_XYZRHW` to avoid mix-ups with `D3DFVF_XYZB1` and up.

elishacloud replied:
Good catch.  Fixed in the latest check in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+					{
+						float *pos = (float*) vertex;
+
+						pos[3] = 1.0f;
```
If this code is never going to handle other scenarios than the intended ones, then indeed just remove it.

But if this work is intended to support all possible use-cases, is better to be safe

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+					const DWORD newVertexTypeDesc = (dwVertexTypeDesc & ~D3DFVF_XYZRHW) | D3DFVF_XYZ;
+
+					// Set fixed function vertex type
+					(*d3d9Device)->SetFVF(newVertexTypeDesc);
```
Since here SetFVF overwrites the value that the client last set, other draw calls after this one might still expect the original value to still be active. To support those use-cases, best remember the currently active value before overwriting it here, and restore the value after the draw call below.

elishacloud replied:
Unlike Direct3D9, in Direct3D7 and older every draw call has the FVF as an argument.  So `SetFVF()` will be called before every draw call.  So there is never a need to restore the value.

Note: this is even true of Direct3D8.  It is only starting with Direct3D9 that you would need to restore the value of FVF.

dkollmann committed:
Fixed incorrect game camera position and orientation. Added option to disable lighting. Commented out unneeded XYZW code.

dkollmann commented:
I made some fixes. Of course lpD3DMatrix has to be replaced, after getting the game camera properties from it.
I added an option to disable the lighting which is needed when DdrawConvertHomogeneousW is true but DdrawConvertHomogeneousToWorld is false. At least for Black & White.

The force push was just an update on the commit comment.

elishacloud commented on ddraw/DebugOverlay.cpp:
```diff
@@ -98,7 +99,7 @@ void DebugOverlay::BeginScene()
 }
 
-void DebugOverlay::EndScene()
+void DebugOverlay::EndScene(const RenderData &RenderData)
```
It looks like `RenderData` is never used in the DebugOverlay.  So there is no need to pass it.

elishacloud replied:
Let's keep this out until we are actually using the data in the overlay.

elishacloud committed:
Fix comments from pull request

elishacloud committed:
Don't override matrix memory, just the pointer

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,15 +339,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+			}
+			else
+			{
+				return D3D_OK;
```
This will basically ignore all other TransformStateTypes.  Won't this break other types, like D3DTS_PROJECTION and D3DTS_WORLD?

dkollmann replied:
The idea is that all matrices are overriden.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2453,11 +2540,99 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
+			// Handle PositionT
+			if (Config.DdrawConvertHomogeneousW && (dwVertexTypeDesc & 0x0E) == D3DFVF_XYZRHW)
+			{
+				D3DFVF_XYZB1;
```
Stray symbol

dkollmann replied:
Oh okay, I have to check this looks like it should not be there.

elishacloud replied:
Fixed in the latest check-in.

elishacloud committed:
Fix disable lighting and a couple typos

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -2050,6 +2131,12 @@ HRESULT m_IDirect3DDeviceX::SetRenderState(D3DRENDERSTATETYPE dwRenderStateType,
+		case D3DRS_LIGHTING:
+			if (Config.DdrawDisableLighting)
+			{
+				dwRenderState = FALSE;
```
Fixed in the latest check-in.

elishacloud commented:
For some reason when I add this `DdrawConvertHomogeneousW = 1` line to the ini file I just get the following:

![image](https://github.com/elishacloud/dxwrapper/assets/25625456/f6f41484-1d73-4cb2-b1d3-ae5a728e4092)

elishacloud committed:
Set transform if not set by game

elishacloud commented:
> For some reason when I add this `DdrawConvertHomogeneousW = 1` line to the ini file I just get the following:

Ok, I figured this out.  You have to set `DdrawDisableLighting = 1` for _Black & White_.

elishacloud commented:
My only last comment is that I prefer to create a struct and add it to the `IDirect3DDeviceX.h` file rather than a whole new class for the `DdrawConvertHomogeneous` settings/matrices.  Since the code is in `IDirect3DDeviceX.cpp` the settings should be in `IDirect3DDeviceX.h`.

Other than that, I think this is about ready to merge.

dkollmann commented:
Sounds good man.

dkollmann commented:
Ah okay, I created the RenderData.h, because it was not possbile to include the device header in the DebugOverlay.cpp file.

elishacloud commented:
> I created the RenderData.h, because it was not possbile to include the device header in the DebugOverlay.cpp file.

`DebugOverlay.h` already has `#include "ddraw.h"`, which includes `IDirect3DDeviceX.h`.  So it is already covered.

elishacloud committed:
Remove RenderData

elishacloud commented:
Now `DdrawConvertHomogeneousW` seems to be working great.  However, I am not able to get `DdrawConvertHomogeneousToWorld` working at all.  I just get a black screen.

PatrickvL commented:
If possible, attach a few good looking screenshots (or even link a video) of the resulting rendering in RTX Remix!

dkollmann commented:
These are my settings

DdrawOverrideWidth         = 1920
DdrawOverrideHeight        = 1080
DdrawOverrideRefreshRate   = 60
DdrawOverrideStencilFormat = 80
DdrawConvertHomogeneousW   = 1
DdrawConvertHomogeneousToWorld = 1
DdrawConvertHomogeneousToWorldUseGameCamera = 1
DdrawConvertHomogeneousToWorldFOV = 30
DdrawConvertHomogeneousToWorldNearPlane = 1
DdrawConvertHomogeneousToWorldFarPlane = 1000
DdrawConvertHomogeneousToWorldDepthOffset = 0.001
DdrawEnableMouseHook       = 0
DdrawDisableLighting       = 1

dkollmann commented:
On my PC with the previous setings...

No RTX, the water plane which should be rendered first is just rendered alogn with the rest of the geometry.
https://1drv.ms/v/s!As0-jjvsSN2xgft8Y6R-1Zsvta99uQ?e=453wxe

RTX, I ignore the wwater plane, so it is not rendered, since it does not serve the purpose of a raytraced surface anyway, since it is a plane facing the camera.
https://1drv.ms/v/s!As0-jjvsSN2xgft9BREiXiT7RT6VAg?e=3t0yky

elishacloud committed:
Fix default settings

elishacloud commented:
Ok, thanks!  I figured it out.  With the latest check-in this all you should need in your ini file.  The rest should be set by default or auto-detected:

```ini
[Compatibility]
Dd7to9                     = 1

[Dd7to9]
DdrawEnableMouseHook       = 0
DdrawDisableLighting       = 1
DdrawConvertHomogeneousW   = 1
DdrawConvertHomogeneousToWorld              = 1
DdrawConvertHomogeneousToWorldUseGameCamera = 1
```

elishacloud commented:
With these settings and the latest build I am not seeing any issues.  Try this build with only the included settings.

Latest build:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/11502785/dxwrapper.zip)

dkollmann commented:
Hmmm it works too well. Something seems off. Let me merge those changes into my branch and let me check.

RTX version crashes, with no explicit stencil format, but that is fine.
DdrawOverrideStencilFormat = 80

dkollmann commented:
Looks correct in blender though.
![image](https://github.com/elishacloud/dxwrapper/assets/936152/785d2c96-3496-4a60-95b5-30828a91a75e)

elishacloud commented:
> RTX version crashes, with no explicit stencil format, but that is fine.
> DdrawOverrideStencilFormat = 80

Sounds like a bug in RTX.  The stencil format is set later.  Maybe we should default to setting a stencil and then let the game override that, if needed?

I defaulted to no stencil because 2D DirectDraw games don't need it.  But I don't think it will hurt to have a default one set.  I prefer having less options and letting dxwrapper figure out things for the user.

Try this one and see if this has the same issue with RTX:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/11502902/dxwrapper.zip)

dkollmann commented:
I would not worry. To use RTX, you have to install all this as a mod anyway, so you also get the correct overrides and new assets.

I want to replace the water shader as well as the terrain shader. And add lights to the spells and so on.

dkollmann committed:
Fixed that DdrawConvertHomogeneousToWorldUseGameCamera did not actually use the game camera transform. Simplified the opposite case.

dkollmann commented:
Okay I found the issue. But still, not using DdrawConvertHomogeneousToWorldUseGameCamera currently yields the better results.

Maybe it would be better to just move the light, instead of transforming the geometry into insane values which then generate Z fighting. For the raytracing it should make no difference and the geometry we see is only the visible area anyway, so moving it around the world has no benefit.

dkollmann commented:
> Try this one and see if this has the same issue with RTX: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/11502902/dxwrapper.zip)

Still crashes. Sorry. But as mentioned it does not matter, the user will get an installer anyway, and that will also install the dxwrapper.ini.

elishacloud commented:
> Still crashes. Sorry. [...] it does not matter, the user will get an installer anyway [...] they just relased RTX Remix 0.2.0 like a week ago. It is still quite an experimental tool.

Ok, no problem.  Then I will not worry about it right now.

> not using DdrawConvertHomogeneousToWorldUseGameCamera currently yields the better results.

Ok.  What's the purpose of DdrawConvertHomogeneousToWorldUseGameCamera?

> Maybe it would be better to just move the light, instead of transforming the geometry into insane values which then generate Z fighting.

I can't reproduce this issue.  I'll wait until you have this resolved before merging the code.

dkollmann commented:
> Ok. What's the purpose of DdrawConvertHomogeneousToWorldUseGameCamera?

The idea is to transform the geometry back to its "original" world position, so it is correctly placed relative to other world geometry. So when I remove the 2D sky and replace it with a sky dome, so the raytracing has something it can actually hit, the game geometry will be correctly placed.

But it will probably be better to transform the sky dome instead into the homogenous space.

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						}
+
+						// Override original matrix pointer
+						lpD3DMatrix = &view;
```
It looks like this line is in there three times.  We can remove two of them.

elishacloud replied:
Ok, this is fixed in the latest check in.

elishacloud committed:
Remove unneeded matrix assignment

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,103 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						{
+							// To reconstruct the 3D world, we need to know where the camera is and where it is looking
+							position = DirectX::XMVectorSet(lpD3DMatrix->_41, lpD3DMatrix->_42, lpD3DMatrix->_43, lpD3DMatrix->_44);
+							direction = DirectX::XMVectorSet(lpD3DMatrix->_31, lpD3DMatrix->_32, lpD3DMatrix->_33, lpD3DMatrix->_34);
```
Oh, I see why this is broken.  lpD3DMatrix is overridden above this.

elishacloud replied:
Ok, this is fixed in the latest check in.

elishacloud committed:
Fix DdrawConvertHomogeneousToWorldUseGameCamera

elishacloud commented:
> But still, not using DdrawConvertHomogeneousToWorldUseGameCamera

I think I fixed this issue.  It should now be using the Game Camera.  See check-in here:  44d43381e096a1fbff974f593ef68abc4d5c7efd

elishacloud commented:
> I can't reproduce this issue.

Now I can see the problem with the DdrawConvertHomogeneousToWorldUseGameCamera.  It works fine as long as this option is disabled.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+					view._44 = 1.0f;
+
+					// Backup pointer before overriding original matrix pointer
+					LPD3DMATRIX lpOriginalD3DMatrix = lpD3DMatrix;
```
This intermediate variable can be avoided if the `lpD3DMatrix` overwrite is moved near the end of this statement block.
That way, here below code uses `lpOriginalD3DMatrix`, it could still use `lpD3DMatrix`.
The only other thing that would have to change, is the one occurrence were below code currently uses `lpD3DMatrix`, it should use `view` instead.

elishacloud replied:
Good idea.  Done in the latest check-in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						DirectX::XMStoreFloat4x4((DirectX::XMFLOAT4X4*)&ConvertHomogeneous.ToWorld_ViewMatrix, viewMatrix);
+
+						// Store the view inverse matrix of the game, so we can transform the geometry with it
+						DirectX::XMMATRIX toViewSpace = DirectX::XMLoadFloat4x4((DirectX::XMFLOAT4X4*)lpD3DMatrix);
```
This could use `&view` instead of `lpD3DMatrix`

elishacloud replied:
Done in the latest check-in.

PatrickvL commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+
+					// Backup pointer before overriding original matrix pointer
+					LPD3DMATRIX lpOriginalD3DMatrix = lpD3DMatrix;
+					lpD3DMatrix = &view;
```
This assignment could be moved lower (see the other remarks on that)

elishacloud replied:
Done in the latest check-in.

elishacloud committed:
Move matrix override code to the end

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						DirectX::XMMATRIX depthoffset = DirectX::XMMatrixTranslation(0.0f, 0.0f, Config.DdrawConvertHomogeneousToWorldDepthOffset);
+
+						ConvertHomogeneous.ToWorld_ViewMatrixInverse = DirectX::XMMatrixMultiply(depthoffset, DirectX::XMMatrixMultiply(toViewSpace, vpinv));
+					}
```
Done in the latest check-in.

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,96 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						if (Config.DdrawConvertHomogeneousToWorldUseGameCamera)
+						{
+							// To reconstruct the 3D world, we need to know where the camera is and where it is looking
+							position = DirectX::XMVectorSet(lpOriginalD3DMatrix->_41, lpOriginalD3DMatrix->_42, lpOriginalD3DMatrix->_43, lpOriginalD3DMatrix->_44);
```
Done in the latest check-in.

elishacloud commented:
Shouldn't [these lines](https://github.com/elishacloud/dxwrapper/pull/199/files#diff-67af21c8c33ff48477ba85b7d7d8b252ac63d62790166c8d610dafc48f14c4caR380-R381) be changed? FROM:

```c++
position = DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 0.0f);
direction = DirectX::XMVectorSet(0.0f, 0.0f, 1.0f, 0.0f);
```

TO:

```c++
position = DirectX::XMVectorSet(view._41, view._42, view._43, view._44);
direction = DirectX::XMVectorSet(view._31, view._32, view._33, view._34);
```

elishacloud commented on ddraw/IDirect3DDeviceX.cpp:
```diff
@@ -338,14 +338,95 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
+						}
+
+						// Store the original matrix so it can be restored
+						ConvertHomogeneous.ToWorld_ViewMatrixOriginal = view;
```
This line is not storing the original matrix.  Should this be changed to:

```c+
ConvertHomogeneous.ToWorld_ViewMatrixOriginal = *lpD3DMatrix;
```

dkollmann replied:
No because the original geometry still needs to be transformed so they work with the expected space.

The original view matrix is useless to transform geometry. It only exists to transform the lights.

This matrix is needed to make geometry look correctly when D3DFVF_XYZRHW is changed to D3DFVF_XYZW.

elishacloud replied:
Ok, I see.

elishacloud commented:
> The result of XMMatrixLookToLH should stay the same.

Are you sure?  The `position` sent would be `DirectX::XMVectorSet(-1.0f, 1.0f, 0.0f, 1.0f);` rather than `DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 0.0f);`.  Also, if we ever change the `view` later than having it set by `view` means it gets updated also.  No need to manually change it later.

elishacloud commented:
It seems like there may be more than just something wrong with `DdrawConvertHomogeneousToWorldUseGameCamera`.  I am seeing an issue in [_Dark Reign 2_](https://www.gog.com/en/game/dark_reign_2) when just using `DdrawConvertHomogeneousToWorld`.  The sky seems stretched

With `DdrawConvertHomogeneousToWorld` enable:

![image](https://github.com/elishacloud/dxwrapper/assets/25625456/12eaf010-7b4e-4870-a981-bb438e250106)

How it is supposed to look:

![image](https://github.com/elishacloud/dxwrapper/assets/25625456/b06d405c-f2b6-426d-9e66-9443f4d735ef)

I can also see it even when just `DdrawConvertHomogeneousW` is enabled, but it is not quite as pronounced.

dkollmann commented:
Okay, here are the things are are wrong here and why the view has a position of zero.

- "lpD3DMatrix = &view;" is set after the whole "if (Config.DdrawConvertHomogeneousToWorld)" block.
- toViewSpace should be "DirectX::XMMATRIX toViewSpace = DirectX::XMLoadFloat4x4((DirectX::XMFLOAT4X4*)&view);"

The result will be the same, but the code should be a bit clearer.

So what happens is...

- With the inverse matrix, "view" is applied. So the geometry is centered in front of the camera.
- Then the inverse view projection is added, so the geometry is transformed into world space.
- From there, the view matrix does not need to contain "view", since the error was already corrected by the inverse matrix.

I hope this all makes sense and is correct. Please take a look.

dkollmann commented:
Not sure if it is related, but I noticed something curious with the reconstructed world psace in Black & White.

![image](https://github.com/elishacloud/dxwrapper/assets/936152/94336746-fbb9-4b8f-a55e-468aa4ec1ce7)

The whole world is rotated by 24.4° degreees. But the thing is, if you look at the trees at the bottom left, they are straight, which implies that this is correct. I wonder why this is. Is this due to the angle of the reflection? In game, the camera is completely horizontal. I checked the original view matrix.

dkollmann commented:
Okay this seems to be a Black & White thing. No matter how I look att he view matrix, the angle only seems to be half of the actual angle. MAybe this is related to faking bent normals.

elishacloud committed:
Fix DdrawConvertHomogeneousToWorldUseGameCamera

elishacloud commented:
@dkollmann, sorry for the delay in responding.  I updated the code to match your branch.  But I am wondering if there is an issue with `DdrawConvertHomogeneousW` which is getting exacerbated in `DdrawConvertHomogeneousToWorld`.

Could there be an issue because we are losing data in the `DrawIndexedPrimitive()` when we overwrite data in `lpVertices`?

The reason I ask is because I am still seeing distortion in the sky in Dark Reign 2 whenever `DdrawConvertHomogeneousToWorld` is enabled, even if `DdrawConvertHomogeneousToWorldUseGameCamera` is disabled.  I also see distortion even if just `DdrawConvertHomogeneousW` is enabled, though the distortion is much smaller in this last case.

dkollmann commented:
Hey, I just wanted to say that I am still on this, but I changed my job and am moving to another country, so lately I am super busy.

elishacloud commented:
@dkollmann, no problem.  Take your time.  I understand.  Real life has to come first.  Thanks for letting me know.

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix merge conflict with main

elishacloud committed:
Minor update

elishacloud commented:
Because there are a lot of things added since this PR was created, I updated it to keep it in-sync with the master branch.

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix formatting

elishacloud committed:
Fix defaults

elishacloud committed:
Merge remote-tracking branch 'origin/master' into pr/199

elishacloud committed:
Fix PR to work with master branch changes

elishacloud committed:
Merge remote-tracking branch 'origin/master' into pr/199

elishacloud committed:
Update to make it match master branch

elishacloud commented:
Keeping PR in-sync with master branch.

elishacloud committed:
Merge remote-tracking branch 'origin/master' into pr/199

elishacloud committed:
Merge remote-tracking branch 'origin/master' into pr/199

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Include DirectXMath

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix issue with merging master branch

elishacloud committed:
Merge remote-tracking branch 'origin/master' into pr/199

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix build merge issues

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix build issue

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix build issue

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Fix build error

elishacloud committed:
Merge branch 'master' into reverse_xyzrhw_new

elishacloud committed:
Remove duplicate DdrawForceMipMapAutoGen variable
